### PR TITLE
Prefer statements to expressions in the Immediate Window

### PR DESCRIPTION
**Pull request by amcasey**
As in the native expression evaluators.

### CODE REVIEW

**Comment by amcasey** [comment]
FYI @jaredpar @ManishJayaswal @cston @KevinH-MS @tmat

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 1]
```diff
@@ -317,26 +317,31 @@ private static MethodSymbol GetSynthesizedMethod(CommonPEModuleBuilder moduleBui
             }
             else
             {
+                formatSpecifiers = null;
+
                 // Try to parse as an expression. If that fails, parse as a statement.
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Fix comment

**Comment by KevinH-MS** [comment]
LGTM :+1: 

Per VB, I don't think we should change anything for the sake of consistency if we can't demonstrate a scenario that's broken (or perhaps I'm misreading the warning note).

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 2]
```diff
@@ -317,26 +317,31 @@ private static MethodSymbol GetSynthesizedMethod(CommonPEModuleBuilder moduleBui
...
+                    return statementSyntax;
+                }
+
                 ReadOnlyCollection<string> exprFormatSpecifiers;
-                CSharpSyntaxNode syntax = expr.ParseExpression(exprDiagnostics, allowFormatSpecifiers: true, formatSpecifiers: out exprFormatSpecifiers);
-                Debug.Assert((syntax == null) || !exprDiagnostics.HasAnyErrors());
-                exprDiagnostics.Free();
-                if (syntax != null)
+                var expressionSyntax = expr.ParseExpression(diagnostics, allowFormatSpecifiers: true, formatSpecifiers: out exprFormatSpecifiers);
+                if (expressionSyntax != null)
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Is the 'if' still needed or could this be "return expr.ParseExpression(..., out formatSpecifiers);"?

**Comment by cston** [comment]
LGTM

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs (hunk 1 above)
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Done.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs (hunk 2 above)
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: It had the effect of keeping the format specifiers null, but that didn't accomplish anything anyway.  Simplified.

**Comment by amcasey** [comment]
@KevinH-MS While I believe it would be worthwhile to make VB consistent with both Roslyn C# and native VB, it does not appear to have a corresponding code path.  I still plan to add comparable regression tests though.

**Comment by KevinH-MS** [comment]
Sounds good (per just adding tests to VB)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 3]
```diff
@@ -311,33 +311,29 @@ private static MethodSymbol GetSynthesizedMethod(CommonPEModuleBuilder moduleBui
...
-                if (syntax != null)
-                {
-                    Debug.Assert(!diagnostics.HasAnyErrors());
-                    formatSpecifiers = exprFormatSpecifiers;
-                    return syntax;
-                }
-                formatSpecifiers = null;
-                syntax = expr.ParseStatement(diagnostics);
-                if ((syntax != null) && (syntax.Kind() != SyntaxKind.LocalDeclarationStatement))
+                // Try to parse as a statement. If that fails, parse as an expression.
```
Comment (ManishJayaswal) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Why did we previously try to parse it as an expression first even when treatAsExpression was false? Are there any cases where that was helpful?

Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: I'll try to dig that out of version control, but I suspect the reason is that we initially lacked support for statements and we wanted to minimize the change the then-current behavior.

Comment (jaredpar) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: @ManishJayaswal for context the native C# EE ignored this flag and made its own decisions about whether to parse as expressions or statements.  It's possible that logic was simply copied over.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 4]
```diff
@@ -311,33 +311,29 @@ private static MethodSymbol GetSynthesizedMethod(CommonPEModuleBuilder moduleBui
...
+                        return statementSyntax;
+                    }
+
                     diagnostics.Add(ErrorCode.ERR_ExpressionOrDeclarationExpected, Location.None);
                     return null;
                 }
-                return syntax;
             }
+
+            return expr.ParseExpression(diagnostics, allowFormatSpecifiers: true, formatSpecifiers: out formatSpecifiers);
```
Comment (jaredpar) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: The native EE attempted three types of parsing in the following order:

1. statements
2. expressions
3. type names

Are we consciously only replicating the first two here?

Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: I believe ParseExpression handles types, but I can confirm.

Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: It does not, but @cston reminded me that we consciously decided not to support type expressions.

**Pull request merged**